### DISCUSSION TITLE
Out of Memory Error and Incorrect Matches in Large Dataset Processing. Help is highly appreciated

### DISCUSSION THREAD

**Discussion by SR-splink** [moj-analytical-services/splink#2573] [Q&A]
Hello Splink Team/Community,
I’m using Splink to implement MPI matching solution for datasets with millions of records. While Splink has been a great tool for probabilistic and deterministic matching, I am facing challenges with processing large datasets and ensuring the accuracy of matches and any help is highly appreciated. currently i am using a fake dataset with 70% overlap between the two inputs and includes 40% errors in spelling. Below are the issues I’m encountering:
1.When processing datasets with ~5 million records in df1 and ~3 million records in df2, the script fails with an "Out of Memory" error , I am running it on my local currently and was able to run the script earlier however dataframes with 1 million records are still processed within couple of mins. I have tried the following with little to no use 
Adjusted blocking rules to reduce the comparison space.
Removed exact matches from probabilistic matching to minimize workload.
Tried chunking datasets but encountered issues where records that should match were distributed across different chunks.
2. Records with the same First_Name and Last_Name but different Birthdate are being matched as valid pairs when blocked on names  even though im using exact match comparison for birthdate and blocking on names . If I change the blocking to birthdate then vice versa(records with same dob but different names matches (if i add first_name in the blocking rule along with birthdate then different lastname matches ) happens even though theres JArowinklerthreshold at place for Names.
I have tried the following 
 "blocking_rules": [
    {"fields": ["Birthdate", "First_Name", "Last_Name"]},
    {"fields": ["First_Name", "Last_Name"]},
    {"fields": ["Birthdate", "Last_Name"]}
]
 "blocking_rules": [
            {"fields" : ["First_Name","BirthDate"]},
            {"fields" : ["BirthDate","Last_Name"]}      
        ],
 "blocking_rules": [
            {"fields" : [BirthDate"]},
            {"fields" : ["First_Name",""Last_Name"]}
        ],
        
        
    basic config file  is as below
    "threshold": 0.95,
    "unique_id_column_name" : "unique_id",
    "comparison_fields" : [
        {
            "type" : "ExactMatch",
            "col_name" : "Birthdate"
           
        },
        {
            "type" : "JaroAtThresholds",
            "col_name" : "First_Name",
            "score_threshold_or_thresholds":[0.95,0.90]
        },
        {
            "type": "JaroAtThresholds",
            "col_name" : "Last_Name",
            "score_threshold_or_thresholds": [0.95,0.90]
        },
        {
            "type"  : "ForenameSurnameComparison",
            "forename_col_name" : "First_Name",
            "surname_col_name" : "Last_Name",
            "jaro_winkler_thresholds": [0.85, 0.75],
            "forename_surname_concat_col_name": "Full_Name"
         
        }
    ],

logic:
 for blocking_rule in settings.blocking_rules_to_generate_predictions:
            linker.training.estimate_parameters_using_expectation_maximisation(blocking_rule)
        # Predict matches with probabilities
        results = linker.inference.predict(threshold_match_probability=0.6)

**Comment by RobinL** [comment]
I would start by analysing the blocking rules to see which ones create the most comparisons using:
https://moj-analytical-services.github.io/splink/demos/tutorials/03_Blocking.html

How many comparisons are you creating and what the spec of your computer?

**Comment by SR-splink** [comment]
Hi , Thank you for your prompt response to help. I could resolve the out of memory error. The problem I am facing right now is more with blocking rules and comparisons. By blocking the data on Birthdate and Firstname fields the algorithm is not comparing the last name field accurately and giving random results with same birthdate and first name but different last name as a match. I have attached the comparisons in the  earlier post.  Please help!

**Reply by RobinL** [reply]
You appear to be including both a ForenameSurnameComparison , and a first name and surname comparisons separately.  This will cause double counting.  Each field should be included in only one comparison

**Reply by SR-splink** [reply]
I was planning on using ForenameSurnameComparison for swapped values between Firstname and surname and then Jarowinkler distance for first name and lastname. Also for the count analysis - I am getting same number of comparisons pre filter and post filter 

{'number_of_comparisons_generated_pre_filter_conditions': 1495958, 'number_of_comparisons_to_be_scored_post_filter_conditions': 1495958, 'filter_conditions_identified': '', 'equi_join_conditions_identified': 'l."First_Name" = r."First_Name" AND l."Birthdate" = r."Birthdate"', 'link_type_join_condition': 'where l."source_dataset" || \'-__-\' || l."unique_id" < r."source_dataset" || \'-__-\' || r."unique_id" and l."source_dataset" != r."source_dataset"'}
<splink.internals.blocking_rule_library.And object at 0x0000019266982B10>
{'number_of_comparisons_generated_pre_filter_conditions': 1151530, 'number_of_comparisons_to_be_scored_post_filter_conditions': 1151530, 'filter_conditions_identified': '', 'equi_join_conditions_identified': 'l."Birthdate" = r."Birthdate" AND l."Last_Name" = r."Last_Name"', 'link_type_join_condition': 'where l."source_dataset" || \'-__-\' || l."unique_id" < r."source_dataset" || \'-__-\' || r."unique_id" and l."source_dataset" != r."source_dataset"'}
